src/lib/FreeC/Pass/TypeInferencePass.hs

```diff
@@ -666,6 +666,22 @@ annotateExprWith' (IR.Lambda srcSpan args expr _) resType =
     addTypeEquation srcSpan funcType resType
     return (IR.Lambda srcSpan args' expr' (makeExprType resType))
 
+
+annotateExprWith' (IR.Let srcSpan binds expr _) resType =
+  withLocalTypeAssumption $ do
+    retType  <- liftConverter freshTypeVar
+    binds'   <- mapM annotateBind binds
+    expr'    <- annotateExprWith expr retType
+    return (IR.Let srcSpan binds' expr' (makeExprType resType))
+     where
+       annotateBind :: IR.Bind -> TypeInference IR.Bind
+       annotateBind (IR.Bind bindSrcSpan bindVarPat bindExpr) = do
+           bindVarPat' <- annotateVarPat bindVarPat
+           bindType    <- liftConverter freshTypeVar
+           bindExpr'   <- annotateExprWith bindExpr bindType
+           addTypeEquation srcSpan (fromJust $ IR.varPatType bindVarPat') bindType
+           return (IR.Bind bindSrcSpan bindVarPat' bindExpr')
+
```

Comment:
This implementation does not infer the type of polymorphic `let` bindings correctly. The following is allowed in Haskell for example.

```haskell
let l = length in l [1,2,3] + l [True,False]
```

I expect polymorphic `let` bindings to be needed for the translation of guards. However, to support them we would have to add type arguments to `let` bindings as shown below. And probably also have to change a lot of code in the `TypeInferencePass`.

```haskell
let l @a = length @a in l @Integer [1,2,3] + l @Bool [True,False]
```

Thus, I'd say that we should ignore this for now. However, there should be documentation and/or a ticket for the missing feature.

Reply:
@macmaccamcam Can you open a ticket and resolve the conversation afterwards?

Reply:
Yes, i was just about to do that.

Reply:
See #157.